### PR TITLE
Fix the coverage import for .razor files

### LINKED ISSUE #7990: Fix the coverage import for .razor files
### Description

CoverageReportImportSensor is throwing an exception in some cases. Please check: https://community.sonarsource.com/t/sonar-cloud-analyze-broken-in-azure-dev-ops/98804

### PR DESCRIPTION
**Pull request by cristian-ambrosini-sonarsource**
Fixes #7990

### CODE REVIEW

File: sonar-dotnet-shared-library/src/test/java/org/sonar/plugins/dotnet/tests/CoverageReportImportSensorTest.java [hunk 1]
```diff
@@ -258,6 +258,22 @@ private SensorContextTester computeCoverageMeasures(boolean isIntegrationTest) {
     return context;
   }
 
+  @Test
```
Comment (costin-zaharia-sonarsource) [inline sonar-dotnet-shared-library/src/test/java/org/sonar/plugins/dotnet/tests/CoverageReportImportSensorTest.java:261]: The test should be added on top of the private methods.

File: sonar-dotnet-shared-library/src/test/java/org/sonar/plugins/dotnet/tests/CoverageReportImportSensorTest.java [hunk 2]
```diff
@@ -258,6 +258,22 @@ private SensorContextTester computeCoverageMeasures(boolean isIntegrationTest) {
...
+  public void computeCoverageLoggingOutOfRange() throws IOException {
```
Comment (costin-zaharia-sonarsource) [inline sonar-dotnet-shared-library/src/test/java/org/sonar/plugins/dotnet/tests/CoverageReportImportSensorTest.java:262]: It doesn't need to throw the `IOException`.

File: sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/CoverageReportImportSensor.java [hunk 3]
```diff
@@ -137,7 +137,13 @@ private static void analyzeCoverage(SensorContext context, Coverage coverage, Fi
     for (Map.Entry<Integer, Integer> entry : coverage.hits(filePath).entrySet()) {
       LOG.trace("Found entry with key '{}' and value '{}'.", entry.getKey(), entry.getValue());
       fileHasCoverage = true;
-      newCoverage.lineHits(entry.getKey(), entry.getValue());
+      var line = entry.getKey();
+      if (line > 0 && line <= inputFile.lines()){
+        newCoverage.lineHits(line, entry.getValue());
+      }
+      else {
+        LOG.trace("Line {} is out of range in the file '{}' (lines: {})", line, inputFile, inputFile.lines());
```
[the remark this answers is not in the document]
Comment (cristian-ambrosini-sonarsource) [reply sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/CoverageReportImportSensor.java]: I just changed them to warning, is it too much?

File: sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/CoverageReportImportSensor.java [hunk 4]
```diff
@@ -134,10 +134,20 @@ private static void analyzeCoverage(SensorContext context, Coverage coverage, Fi
...
+      var line = entry.getKey();
+      if (line <= inputFile.lines()){
+        newCoverage.lineHits(line, entry.getValue());
+      } else {
+        coverageImportError = true;
+        LOG.debug("Coverage import: Line {} is out of range in the file '{}' (lines: {})", line, inputFile, inputFile.lines());
+      }
+    }
+    if (coverageImportError){
+      LOG.warn("Verify or raise an issue with the coverage tool being used, as the imported coverage file has reported lines that are out of range for certain file(s).");
```
Comment (costin-zaharia-sonarsource) [inline sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/CoverageReportImportSensor.java:150]: Should we mention that more details are available in the debug logs?

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**